File: SocialJournal/SocialJournal/EntryViewController.m
//
//  EntryViewController.m
//  SocialJournal
//
//  Created by James Garcia on 11/9/14.
//  Copyright (c) 2014 UH. All rights reserved.
//

#import "EntryViewController.h"
#import "CommentTableViewCell.h"

@interface EntryViewController ()
@property (weak, nonatomic) IBOutlet UIView *whiteBackgroundView;
@property UITextField *commentAlertTextField;
@property (weak, nonatomic) IBOutlet UITextField *titleTextField;
@property (weak, nonatomic) IBOutlet UITextField *tagsTextField;
@property (weak, nonatomic) IBOutlet UITextView *contentTextField;
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UILabel *profileNameLabel;
@property (weak, nonatomic) IBOutlet UILabel *labelDayName;
@property (weak, nonatomic) IBOutlet UILabel *labelDay;
@property (weak, nonatomic) IBOutlet UILabel *labelMonth;
@property (weak, nonatomic) IBOutlet UILabel *labelYear;

@end

@implementation EntryViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.profileNameLabel.text = self.username;
    self.whiteBackgroundView.layer.masksToBounds = NO;
    self.whiteBackgroundView.layer.shadowOffset = CGSizeMake(-15, 0);
    self.whiteBackgroundView.layer.shadowRadius = 5;
    self.whiteBackgroundView.layer.shadowOpacity = 0.7;
    self.titleTextField.text = self.entry[@"title"];
    self.contentTextField.text = self.entry[@"entry"];
    [self assignDate];
}

- (void) assignDate{
    
    NSDateFormatter *dateFormatter = [[NSDateFormatter alloc] init];
    
    [dateFormatter setDateFormat:@"EEEE"];
    self.labelDayName.text = [dateFormatter stringFromDate:[self.entry createdAt]];
    
    [dateFormatter setDateFormat:@"dd"];
    self.labelDay.text = [dateFormatter stringFromDate:[self.entry createdAt]];
    
    [dateFormatter setDateFormat:@"MMMM"];
    self.labelMonth.text = [dateFormatter stringFromDate:[self.entry createdAt]];
    
    [dateFormatter setDateFormat:@"YYYY"];
    self.labelYear.text = [dateFormatter stringFromDate:[self.entry createdAt]];
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)heartButtonClicked:(UIButton *)sender {
    if ([[sender.imageView accessibilityIdentifier] isEqualToString:@"HeartWhite"]) {
        UIImageView *newImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"HeartRed"]];
        [sender setImage:newImageView.image forState:UIControlStateNormal];
    }else{
        UIImageView *newImageView = [[UIImageView alloc] initWithImage:[UIImage imageNamed:@"HeartWhite"]];
        [sender setImage:newImageView.image forState:UIControlStateNormal];
    }
}

- (IBAction)commentButtonClicked:(id)sender {
    UIAlertView * alert = [[UIAlertView alloc] initWithTitle:@"Submit new comment" message:nil delegate:self cancelButtonTitle:@"Cancel" otherButtonTitles:@"Submit", nil];
    alert.alertViewStyle = UIAlertViewStylePlainTextInput;
    self.commentAlertTextField = [alert textFieldAtIndex:0];
    self.commentAlertTextField.placeholder = @"Enter your comment...";
    [alert dismissWithClickedButtonIndex:0 animated:YES];
    [alert show];
}

- (void)alertView:(UIAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex {
    if (buttonIndex == 1) {
        PFObject *newComment = [PFObject objectWithClassName:@"Comments"];
        newComment[@"content"] = self.commentAlertTextField.text;
        newComment[@"user"] = [PFUser currentUser];
        newComment[@"entry"] = self.entry;
        [newComment saveEventually];
    }
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.pfComments.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    CommentTableViewCell *cell = (CommentTableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"CommentTableViewCell"];
    if (cell == nil)
    {
        NSArray *xib = [[NSBundle mainBundle] loadNibNamed:@"CommentTableViewCell" owner:self options:nil];
        cell = [xib objectAtIndex:0];
    }
    
    cell.tintColor = [UIColor colorWithRed:(242/255.0) green:(242/255.0) blue:(242/255.0) alpha:1.0];
    
    PFObject *currentCellObject = [self.pfComments objectAtIndex:indexPath.row];
    PFObject *userNameForComment = [self.usersOfComments objectAtIndex:indexPath.row];
    cell.userName.text = userNameForComment[@"username"];
    cell.commentText.text = currentCellObject[@"content"];
    cell.dateAndCoordinates.text = @"Coordinates go here";
    
    return cell;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    return 100;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSLog(@"Comment clicked!");
}

- (IBAction)showCommentsButtonClicked:(UIButton *)sender
{
    [UIView beginAnimations:nil context:NULL];
    [UIView setAnimationDuration:0.3];
    [UIView setAnimationCurve:UIViewAnimationCurveEaseOut];
    if (sender.frame.origin.y > 700) {
        sender.frame = CGRectMake(147, 506, 556, 44);
        [sender setTitle:@"Hide Comments" forState:UIControlStateNormal];
        self.tableView.frame = CGRectMake(147, 548, 556, 220);
    }else{
        sender.frame = CGRectMake(147, 724, 556, 44);
        [sender setTitle:@"Show Comments" forState:UIControlStateNormal];
        self.tableView.frame = CGRectMake(147, 771, 556, 220);
    }
    
    [UIView commitAnimations];
}


@end

File: SocialJournalSwift/SocialJournal/SocialJournal/EntryViewController.swift
//
//  EntryViewController.swift
//  SocialJournal
//
//  Created by James Garcia on 11/19/14.
//  Copyright (c) 2014 UH. All rights reserved.
//

import UIKit

class EntryViewController: UIViewController, UITableViewDataSource, UITableViewDelegate {
    var button: HamburgerButton! = nil
    
    @IBOutlet weak var showCommentsVisualView: UIVisualEffectView!
    var showCommentsToggle = true
    @IBOutlet weak var commentsTable: UITableView!
    
    @IBOutlet weak var deleteButton: UIBarButtonItem!
    @IBOutlet weak var username: UILabel!
    @IBOutlet weak var userProfilePicture: UIImageView!
    @IBOutlet weak var heartLike: UIButton!
    @IBOutlet weak var heartCount: UILabel!
    @IBOutlet weak var postTitle: UILabel!
    @IBOutlet weak var postBody: UITextView!
    @IBOutlet weak var dateWeekday: UILabel!
    @IBOutlet weak var dateDay: UILabel!
    @IBOutlet weak var dateMonth: UILabel!
    @IBOutlet weak var dateYear: UILabel!
    @IBOutlet weak var shareButton: UIButton!
    var entry = PFObject(className: "Entry")
    
    var comments = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        

        self.entry["user"].fetchIfNeeded()
        var user:PFUser = self.entry["user"] as PFUser
        self.username.text = user.username
        
        
        if(self.navigationItem.hidesBackButton){
            self.setupTheHamburgerIcon()
            self.navigationItem.setHidesBackButton(false, animated: true)
        }
        
        if(self.username.text == PFUser.currentUser().username){
            self.deleteButton.enabled = true
            self.deleteButton.tintColor = UIColor.redColor()
            
        }else{
            self.deleteButton.tintColor = UIColor.clearColor()
            self.deleteButton.enabled = false;
        }
        var userImageFile:PFFile? = nil
        
        userImageFile = user["profileImage"] as? PFFile
        if userImageFile != nil {
            userImageFile!.getDataInBackgroundWithBlock {
                (imageData: NSData!, error: NSError!) -> Void in
                if error == nil {
                    self.userProfilePicture.image = UIImage(data:imageData)
                }
            }
        }
        
        self.userProfilePicture = prettifyImage(self.userProfilePicture)
        
        var entryTitle:String = entry["title"] as String!
        var entryText:String = entry["content"] as String!
        
        self.heartCount.text = String(ParseQueries.getHeartCountForEntry(self.entry))
        self.postTitle.text = entryTitle
        self.postBody.text = entryText
        assignDate(self.entry.createdAt)
        
        self.postBody.textColor = UIColor.whiteColor()
        self.postBody.font = UIFont(name: "HelveticaNeue-Light", size: 22)
        
        showCommentsVisualView.layer.borderWidth = 1.0
        showCommentsVisualView.layer.borderColor = UIColor.whiteColor().CGColor
        
        var query = PFQuery(className: "Activity")
        query.whereKey("entry", equalTo: entry)
        query.whereKey("fromUser", equalTo: PFUser.currentUser())
        query.whereKey("type", equalTo: "like")
        var likes = query.findObjects()
        
        if likes.count > 0 {
            self.heartLike.tintColor = UIColor.redColor()
            self.heartLike.setImage(UIImage(named: "HeartRed"), forState: .Normal)
        }
        
        getCommentsForEntry()
        
    }
    
    func prettifyImage(imageViewToModify: UIImageView) -> UIImageView{
        imageViewToModify.layer.cornerRadius = imageViewToModify.frame.size.width / 2;
        imageViewToModify.clipsToBounds = true;
        imageViewToModify.layer.borderWidth = 1.0
        imageViewToModify.layer.borderColor = UIColor.whiteColor().CGColor;
        return imageViewToModify
    }
    
    func getCommentsForEntry() {
        var query = PFQuery(className: "Activity")
        query.whereKey("type", equalTo: "comment")
        query.whereKey("entry", equalTo: self.entry)
        query.findObjectsInBackgroundWithBlock {
            (objects: [AnyObject]!, error: NSError!) -> Void in
            if error == nil {
                self.comments = objects
                self.commentsTable.reloadData()
            } else {
                NSLog("Error: %@ %@", error, error.userInfo!)
            }
        }
    }
    
    func assignDate(date:NSDate) {
        var dateFormatter = NSDateFormatter()
        
        dateFormatter.setLocalizedDateFormatFromTemplate("EEEE")
        self.dateWeekday.text = dateFormatter.stringFromDate(date)
        
        dateFormatter.setLocalizedDateFormatFromTemplate("dd")
        self.dateDay.text = dateFormatter.stringFromDate(date)
        
        dateFormatter.setLocalizedDateFormatFromTemplate("MMMM")
        self.dateMonth.text = dateFormatter.stringFromDate(date)
        
        dateFormatter.setLocalizedDateFormatFromTemplate("YYYY")
        self.dateYear.text = dateFormatter.stringFromDate(date)
    }
    
    @IBAction func touchShare(sender: AnyObject) {
        let firstActivityItem = entry["title"] as String!
        let activityViewController : UIActivityViewController = UIActivityViewController(activityItems: [firstActivityItem], applicationActivities: nil)
        
        activityViewController.popoverPresentationController?.sourceView = self.shareButton
        activityViewController.popoverPresentationController?.sourceRect = CGRectMake(self.shareButton.frame.width/2, 0, 0, 0)
        
        self.presentViewController(activityViewController, animated: true, completion: nil)
        
    }
    
    @IBAction func clickDelete(sender: AnyObject) {
        
        var alert = UIAlertController(title: "DELETE ENTRY", message: "Are you sure you want to delete this entry?", preferredStyle: UIAlertControllerStyle.Alert)
        alert.addAction(UIAlertAction(title: "Yes", style: UIAlertActionStyle.Destructive, handler: {(alert: UIAlertAction!) in (self.deleteEntry())}))
        alert.addAction(UIAlertAction(title: "No", style: UIAlertActionStyle.Cancel, handler: {(alert: UIAlertAction!) in println("No Delete")}))
        
        self.presentViewController(alert, animated: true, completion: nil)
    }
    
    func deleteEntry(){
        self.entry.deleteEventually()
        self.navigationController?.popViewControllerAnimated(true)
    }
    
    
    @IBAction func heartPost(sender: AnyObject) {
        
        var newHeartCount = 0
        var query = PFQuery(className: "Activity")
        query.whereKey("fromUser", equalTo: PFUser.currentUser())
        query.whereKey("toUser", equalTo: self.entry["user"] as PFUser)
        query.whereKey("entry", equalTo: self.entry)
        query.findObjectsInBackgroundWithBlock {
            (objects: [AnyObject]!, error: NSError!) -> Void in
            if error == nil {
                if objects.count == 0 {
                    self.heartLike.tintColor = UIColor.redColor()
                    self.heartLike.setImage(UIImage(named: "HeartRed"), forState: .Normal)
                    newHeartCount = self.heartCount.text!.toInt()! + 1
                    self.heartCount.text = String(newHeartCount)
                    self.entry["user"].fetchIfNeededInBackgroundWithBlock {
                        (object: PFObject!, error: NSError!) -> Void in
                        if error == nil {
                            
                            var activity = PFObject(className: "Activity")
                            activity["fromUser"] = PFUser.currentUser()
                            activity["toUser"] = object as PFUser
                            activity["type"] = "like"
                            activity["entry"] = self.entry
                            activity.saveInBackground()
                            
                        } else {
                            NSLog("Error: %@ %@", error, error.userInfo!)
                        }
                    }
                } else {
                    objects[0].deleteInBackground()
                    self.heartLike.tintColor = UIColor.whiteColor()
                    self.heartLike.setImage(UIImage(named: "HeartWhite"), forState: .Normal)
                    newHeartCount = self.heartCount.text!.toInt()! - 1
                    self.heartCount.text = String(newHeartCount)
                }
            } else {
                NSLog("Error: %@ %@", error, error.userInfo!)
            }
        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }    
    
    
    func setupTheHamburgerIcon() {
        self.button = HamburgerButton(frame: CGRectMake(20, 20, 60, 60))
        self.button.addTarget(self, action: "toggle:", forControlEvents:.TouchUpInside)
        self.button.addTarget(self.revealViewController(), action: "revealToggle:", forControlEvents: UIControlEvents.TouchUpInside)
        // self.view.addGestureRecognizer(self.revealViewController().panGestureRecognizer())
        
        self.navigationController?.setNavigationBarHidden(false, animated: true)
        var myCustomBackButtonItem:UIBarButtonItem = UIBarButtonItem(customView: self.button)
        self.navigationItem.leftBarButtonItem  = myCustomBackButtonItem
    }
    
    func toggle(sender: AnyObject!) {
        self.button.showsMenu = !self.button.showsMenu
    }

    @IBAction func commentButtonClicked(sender: UIButton) {
        var currentFrame = showCommentsVisualView.frame
        UIView.beginAnimations(nil, context: nil)
        UIView.setAnimationDuration(0.3)
        UIView.setAnimationCurve(UIViewAnimationCurve.EaseInOut)
        
        if (showCommentsToggle){
            showCommentsVisualView.frame = CGRectMake(currentFrame.origin.x, currentFrame.origin.y-300, currentFrame.size.width, currentFrame.size.height+300)
            sender.layer.backgroundColor = UIColor.blackColor().CGColor
            sender.setTitle("Hide Comments", forState: UIControlState.Normal)
            showCommentsToggle = !showCommentsToggle
        }else{
            showCommentsVisualView.frame = CGRectMake(currentFrame.origin.x, currentFrame.origin.y+300, currentFrame.size.width, currentFrame.size.height-300)
            sender.layer.backgroundColor = UIColor.clearColor().CGColor
            sender.setTitle("Show Comments", forState: UIControlState.Normal)
            showCommentsToggle = !showCommentsToggle
        }
        
        UIView.commitAnimations()
    }
    
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        self.commentsTable.deselectRowAtIndexPath(indexPath, animated: true)
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.comments.count
    }
    
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        var comment:PFObject = self.comments[indexPath.row] as PFObject
        var user:PFUser = comment["fromUser"] as PFUser
        user.fetchIfNeeded()
        
        var cell:CommentCell = tableView.dequeueReusableCellWithIdentifier("commentCell") as CommentCell
        cell.userName.text = user.username
        cell.theComment.text = (comment["content"] as String)
        cell.backgroundColor = UIColor.clearColor()
        
        var userImageFile:PFFile? = user["profileImage"] as? PFFile
        var imageData = userImageFile?.getData()
        if imageData != nil {
            cell.userProfilePicture.image = UIImage(data: imageData!)
        }
        
        
        return cell
    }
    
    func tableView(tableView:UITableView!, heightForRowAtIndexPath indexPath:NSIndexPath)->CGFloat
    {
        return 100
    }
    
    
    @IBAction func locationButtonPressed(sender: AnyObject) {
        if self.entry["location"].latitude != 0.0 && self.entry["location"].longitude != 0.0{
            self.performSegueWithIdentifier("entryToMapView", sender: sender)
        }else{
            let alertController = UIAlertController(title: "Sorry!", message:
                "There is no location attached to this post.", preferredStyle: UIAlertControllerStyle.Alert)
            alertController.addAction(UIAlertAction(title: "Dismiss", style: UIAlertActionStyle.Default,handler: nil))
            
            self.presentViewController(alertController, animated: true, completion: nil)
        }
    }
    
    @IBAction func submitCommentButtonTapped(sender: UIButton) {
        
        let alertController = UIAlertController(title: "Submit new comment", message: nil, preferredStyle: .Alert)

        let submitAction = UIAlertAction(title: "Submit", style: .Default) { (_) in
            var newComment = PFObject(className: "Activity")
            newComment["fromUser"] = PFUser.currentUser()
            newComment["toUser"] = self.entry["user"]
            newComment["type"] = "comment"
            newComment["content"] = (alertController.textFields![0] as UITextField).text
            newComment["entry"] = self.entry
            newComment.saveInBackgroundWithBlock {
                (success: Bool!, error: NSError!) -> Void in
                if (success != nil) {
                    self.getCommentsForEntry()
                } else {
                    NSLog("%@", error)
                }
            }
        }
        submitAction.enabled = false
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .Cancel) { (_) in }
        
        alertController.addTextFieldWithConfigurationHandler { (textField) in
            textField.placeholder = "Your comment..."
            
            NSNotificationCenter.defaultCenter().addObserverForName(UITextFieldTextDidChangeNotification, object: textField, queue: NSOperationQueue.mainQueue()) { (notification) in
                submitAction.enabled = textField.text != ""
            }
        }
        
        alertController.addAction(cancelAction)
        alertController.addAction(submitAction)
        
        self.presentViewController(alertController, animated: true, completion: nil)

    }
    
    
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject!) {
        
        if segue.identifier == "entryToMapView"{
            let vc = segue.destinationViewController as MapViewController
            vc.currentEntry = self.entry as PFObject
        }
    }
    

    
}
